src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::coord::{Coordinator, Message};
+use itertools::Itertools;
+use mz_catalog::memory::objects::{CatalogItem, ClusterVariant, ClusterVariantManaged};
+use mz_controller_types::ClusterId;
+use mz_ore::soft_panic_or_log;
+use mz_sql::catalog::CatalogCluster;
+use mz_sql_parser::ast::ClusterScheduleOptionValue;
+use tracing::warn;
+
+const POLICIES: &[&str] = &[REFRESH_POLICY_NAME];
+
+const REFRESH_POLICY_NAME: &str = "refresh";
+
+impl Coordinator {
+    #[mz_ore::instrument(level = "debug")]
+    /// Call each scheduling policy, gather their decisions about turning clusters on/off, and send
+    /// a message with the result.
+    pub(crate) async fn check_scheduling_policies(&mut self) {
+        // (So far, we have only this one policy.)
+        let refresh_policy_decisions = self.check_refresh_policy().await;
```

Comment:
Oh, hm, it'll be tricky to pull the matview write frontiers from off the main thread, though... but I think no matter what we need to find a way to get the call to `self.get_local_timestamp_oracle()` off the main thread. @jkosh44 or @ParkMyCar can confirm.

Reply:
We should also add counter metrics for how long we're spending in `check_scheduling_policies`, `check_refresh_policy`, and `handle_scheduling_decisions` (or whatever subset of those makes sense). Good to have a birds eye view of how expensive these functions are.

Reply:
> Oh, hm, it'll be tricky to pull the matview write frontiers from off the main thread, though... but I think no matter what we need to find a way to get the call to self.get_local_timestamp_oracle() off the main thread.

How about getting the matview write frontiers on the main thread, passing them to a background task, and then calling `ts_oracle.read_ts()` on the task?

I think just getting the write frontiers of matviews on `SCHEDULE = REFRESH` clusters should be fast. There are other things doing much more complex operations every second, e.g. `advance_timelines` (which calls the rather complicated `update_read_holds` for every index and matview, and also iterates through _all_ catalog objects inside `ids_in_timeline`, not just catalog objects in `SCHEDULE = ON REFRESH` clusters), or `ActiveInstance::update_write_frontiers` / `update_read_capabilities` (which do really complex stuff for every Compute collection).

Also, I'll add metrics for how long we're spending in these functions, and then if it's deemed too much we can increase the interval to something like 10-20 sec.

Reply:
> what do y'all think about bumping the check to 10s or 30s?

I think 10s should be fine. If there is a `REFRESH EVERY '1 day'` matview with a warmup of an hour or maybe tens of minutes, then a 10s variation on when exactly the replica comes on shouldn't cause an issue.

(And in the tests I would manually set the interval to a lower value, so that the tests don't take so much time.)

Reply:
I've moved querying the timestamp oracle to a background task.
0335e1b8b506d0b056fdfaf0cc7657490486ddfe